### PR TITLE
Checkmatefile parser and schema definition, part 1

### PR DESCRIPTION
**Pull request by larsbutler**
Update: <strike>This needs some fixes. Please don't merge yet.</strike> Okay, ready to go.

Partially addresses https://github.com/checkmate/chessboard/issues/7.

This change adds the first iteration of Checkmatefile schema parsing and validation. There are tons of TODOs and FIXMEs in the code to give hints about what needs to be done next. 

In particular, we need to add schema definition and handling for the "long form" of things like `relations` and `options`.

Test coverage is near-perfect: 97% (missing two lines).

### CODE REVIEW

File: chessboard/schema.py [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+    volup.Optional('display-name'): str,
+})
+
+
+#: Schema for `blueprint`
+BLUEPRINT_SCHEMA = volup.Schema({
+    # NOTE(larsbutler): Blueprint authors may be inclined to use a numeric id,
+    # but a string is allowed. Thus, we need to coerce it to a string, in case
+    # the id is an int, in order to avoid an error.
+    volup.Required('id'): Coerce(str),
```
Comment (ziadsawalha) [inline chessboard/schema.py]: In YAML, the number would be converted in unanticipated ways:
```
100 -> "100"
0100 -> "64"
0800 -> "0800"
0x0 -> "0"
```
So I think an error saying “name cannot be an integer” would be safer than coercion.

File: chessboard/tests/test_parser.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        """Test the formatting of error paths."""
+        schema = volup.Schema({
+            volup.Required('blueprint'): volup.Schema({
+                volup.Required('id'): str,
+                volup.Required('name'): str,
+                volup.Optional('description'): str,
+            }),
+        })
+
+        fileobj = six.StringIO("""\
```
Comment (ziadsawalha) [inline chessboard/tests/test_parser.py]: We don't _need_ the `\`. Is there a reason to add it?

File: chessboard/tests/test_parser.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+blueprint:
+    id: def456
+inputs: {}
+""")
```
Comment (ziadsawalha) [inline chessboard/tests/test_parser.py]: If we tab this, would it flow better for reading? The YAML is still valid...
```python
        fileobj = six.StringIO("""
            blueprint:
                id: def456
            inputs: {}
        """)
```

File: chessboard/parser.py [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    :returns:
+        Checkmatefile contents as a `dict`.
+    """
+    contents = fileobj.read()
+    data = yaml.safe_load(contents)
+
+    # Validate the contents per the checkmatefile schema
+    try:
+        # NOTE(larsbutler): Voluptuous schema validators can mutate data, so we
+        # need to return whatever the schema validitor gives us.
```
Comment (ziadsawalha) [inline chessboard/parser.py]: typo: validitor

File: chessboard/parser.py [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+        valid_data = schema(data)
```
Comment (ziadsawalha) [inline chessboard/parser.py]: Why not just `return schema(data)`?

File: chessboard/schema.py [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+    #         interfaces with the same type; this tag is used to differentiate
+    #         them. Expands to `connect-from` in the long form.
+    #         TODO(larsbutler): Why is it `connect-from`, and not `connect-to`?
+    #         It seems backwards.
+    volup.Required('service'): str,
+    volup.Required('interface'): volup.Any('*', *INTERFACE_TYPES),
+
+    # FIXME(larsbutler): Need a better description for all of these.
+    # "source connection point name"
+    volup.Optional('connect-from'): str,
```
Comment (ziadsawalha) [inline chessboard/schema.py]: OK. Should we merge this as is and then work on docs and tweaks on top of it?

Comment (larsbutler) [inline chessboard/schema.py]: I will fix the minor issues you've signalled, but the docs probably will need to wait. I don't have the information or context, and thus am the wrong person to write the docs.

**Comment by larsbutler** [comment]
@ziadsawalha Thanks for the feedback; I got your comments addressed. Anything else?

File: chessboard/parser.py [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+"""
+
+import six
+import voluptuous as volup
+import yaml
+
+from chessboard import schema as cb_schema
+
+
+class MultiValidationError(Exception):
```
Comment (samstav) [inline chessboard/parser.py]: This is neat. What do you think about submitting it upstream to [alecthomas/voluptuous](https://github.com/alecthomas/voluptuous)?

...
update: 

I just looked over the source of voluptuous.py and I was trying to think of a way you could integrate it without completely replacing the MultipleInvalid class, but I'm not sure...

File: chessboard/schema.py [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+    'load-balancer',
+    'mail-relay',
+    'web',
+    'monitoring',
+    'storage',
+    'volume',
+]
+
+
+def DictOf(schema):
```
Comment (samstav) [inline chessboard/schema.py]: I don't think `schema` is defined of imported from anywhere?

Comment (larsbutler) [inline chessboard/schema.py]: `schema` is just a function parameter.

Comment (samstav) [inline chessboard/schema.py]: lol oops. I saw caps and my brain went `class`

File: chessboard/parser.py (hunk 7 above)
Comment (larsbutler) [inline chessboard/parser.py]: Not a bad idea. I'll think about it.

File: chessboard/schema.py (hunk 8 above)
Comment (larsbutler) [inline chessboard/schema.py]: Haha, yeah I figured that's what happened. =)

File: chessboard/schema.py [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+
+    If the type constructor throws a ValueError, the value will be marked as
+    Invalid.
+
+    Shamelessly copied from the voluptuous docs:
+    https://github.com/alecthomas/voluptuous/blob/master/README.md
+    """
+    def f(v):
+        try:
+            return type(v)
```
Comment (samstav) [inline chessboard/schema.py]: I would avoid using `type` as a variable name. We can blame alecthomas since it was in his README I guess?

Comment (larsbutler) [inline chessboard/schema.py]: Yeah, I can change it.

Comment (larsbutler) [inline chessboard/schema.py]: I'll add that to the next schema patch.

File: chessboard/schema.py [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+        except ValueError:
+            raise volup.Invalid(msg or ('expected %s' % type.__name__))
+    return f
+
+
+######################
+# Schema definitions #
+######################
```
Comment (samstav) [inline chessboard/schema.py]: I would vote for adding an explicit `extra=volup.PREVENT_EXTRA` to the usages of `Schema` below.

**Comment by samstav** [comment]
Very nice @larsbutler . Sorry I was late on the review.

File: chessboard/schema.py (hunk 10 above)
Comment (larsbutler) [inline chessboard/schema.py]: Why so? Are you afraid that they're going to change the default?

Comment (samstav) [inline chessboard/schema.py]: Not so much that, I just thought it would help the code document itself.

**Pull request merged**